xapian-applications/omega/clickmodel/simplifieddbn.cc

```diff
@@ -0,0 +1,213 @@
+
+		// Convert click count string to an integer.
+		stringstream ss(clickcount);
+		ss >> click;
+	    } else {
+		clickstring.clear();
+		clicks.push_back(click);
+	    }
+	}
+    }
+    return clicks;
+}
+
+vector<vector<string>>
+SimplifiedDBN::build_sessions(string logfile)
+{
+    ifstream file;
+    file.open(logfile, ios::in);
+
+    string line;
+
+    // Skip the first line.
+    getline(file, line);
+
+    vector<vector<string>> sessions;
+    vector<string> session_entry;
+
+    // Start reading file from the second line.
+    while (getline(file, line)) {
+
+	istringstream ss(line);
+
+	vector<string> row_data;
+
+	while (ss >> std::ws) {
+	    string column_element;
+	    if (ss.peek() == '"') {
+		int pos = ss.tellg();
+		ss.seekg(pos + 1);
+		char ch;
+		while (ss.get(ch)) {
+		    if (ch == '"')
+			break;
+		    column_element += ch;
+		}
+	    } else {
+		getline(ss, column_element, ',');
+	    }
+	    row_data.push_back(column_element);
+	}
+
+	string qid = row_data[0];
+	string query = row_data[1];
+	string docids = row_data[2];
+	string clicks = row_data[4];
+
+	session_entry.push_back(qid);
+	session_entry.push_back(docids);
+	session_entry.push_back(clicks);
+
+	sessions.push_back(session_entry);
+
+	session_entry.clear();
+    }
+
+    file.close();
+    return sessions;
+}
+
+void
+SimplifiedDBN::train(vector<vector<string>> &sessions)
+{
+    map<string, map<string, map<char, array<double, 2>>>> url_rel_fractions;
+
+    for (size_t i = 0; i < sessions.size(); ++i) {
+	string qid = sessions[i][QID];
+
+	vector<string> docids = get_docid_list(sessions);
+
+	vector<int> clicks = get_click_list(sessions);
+
+	int last_clicked_pos = clicks.size() - 1;
+
+	for (size_t j = 0; j < clicks.size(); ++j)
+	    if (clicks[j] != 0)
+		last_clicked_pos = j;
+
+	vector<int> session_clicks;
+	for (int x = 0; x <= last_clicked_pos + 1; ++x)
+	    session_clicks.push_back(clicks[x]);
+
+	for (size_t k = 0; k < session_clicks.size(); ++k) {
+	    if (session_clicks[k] != 0) {
+		url_rel_fractions[qid][docids[k]]['a'][1] += 1;
+		if (int(k) == last_clicked_pos)
+		    url_rel_fractions[qid][docids[k]]['s'][1] += 1;
+		else
+		    url_rel_fractions[qid][docids[k]]['s'][0] += 1;
+	    } else {
+		url_rel_fractions[qid][docids[k]]['a'][0] += 1;
+	    }
+	}
+    }
+
+    for (auto i = url_rel_fractions.begin(); i != url_rel_fractions.end(); ++i) {
+	string qid = i->first;
+	for (auto j = i->second.begin(); j != i->second.end(); ++j) {
+	    auto rel_fractions = j->second;
+	    url_relevances[qid][j->first][PARAM_ATTR_PROB] = rel_fractions['a'][1] /
+							    (rel_fractions['a'][1] +
+							     rel_fractions['a'][0]);
+	    url_relevances[qid][j->first][PARAM_SAT_PROB] = rel_fractions['s'][1] /
+							    (rel_fractions['s'][1] +
+							     rel_fractions['s'][0]);
+	}
+    }
+}
+
+vector<int>
+SimplifiedDBN::get_predicted_relevances(vector<vector<string>> &sessions)
+{
+    vector<int> relevances;
+    double a, s;
```

Comment:
Better to declare variables at the point of initialisation when you can, then there are no undefined values which could be used by mistake.  And you trivially can here - just merge these declarations into the assignments in the loop.